### PR TITLE
Add application metadata to Application audit events

### LINKED ISSUE #8736: Add application name to audit logging when accessing an application
### What
Have Teleport report the application name in the audit logs when accessing an application (APP SESSION DATA) so it is easy to tell which application is being accessed.

### How
Add the application name to the audit log event APP SESSION DATA emitted when an application is accessed by a user

### Why
This will allow users to be able to tell exactly which application was being accessed when multiple applications are on a single server

### PR DESCRIPTION
**Pull request by gabrielcorado**
Closes #8736 .

Changes the events Application audit events to have an `AppMetadata` field. This field contains information about the Application, like its name and public address. Following the changes made to the events:

`AppSessionStart`:
```diff
{
  "addr.remote": "[::1]:63726",
+ "app_name": "application-name",
+ "app_public_addr": "application-address",
  "cluster_name": "localhost",
  "code": "T2007I",
  "ei": 0,
  "event": "app.session.start",
  "namespace": "default",
  "public_addr": "application-address",
  "server_id": "b8751b41-701a-4ce5-a8a0-3d55a884bcb3",
  "sid": "937ee821-7e70-4eb9-a946-8bf235e49fde",
  "time": "2021-11-10T19:42:40.503Z",
  "uid": "10796896-729f-4f35-89a3-cef3f840d393",
  "user": "user"
}
```

`AppSessionChunk`:
```diff
{
+ "app_name": "application-name",
+ "app_public_addr": "application-address",
  "cluster_name": "localhost",
  "code": "T2008I",
  "ei": 0,
  "event": "app.session.chunk",
  "namespace": "default",
  "server_id": "b8751b41-701a-4ce5-a8a0-3d55a884bcb3",
  "session_chunk_id": "b9aeb9c7-8838-4559-be37-2d4a3cd6c55a",
  "sid": "937ee821-7e70-4eb9-a946-8bf235e49fde",
  "time": "2021-11-10T19:42:40.733Z",
  "uid": "9aa3ee2c-4b6d-4595-bf40-b3bcb2621e8f",
  "user": "user"
}
```

`AppSessionRequest`:
```diff
{
  "ei": 5,
  "event": "app.session.request",
  "uid": "d1461ee4-69a3-4f20-891d-66ae6cc053d6",
  "code": "T2009I",
  "time": "2021-11-10T18:26:14.977Z",
  "cluster_name": "localhost",
  "status_code": 200,
  "path": "/",
  "raw_query": "",
  "method": "GET",
+ "app_public_addr": "application-address",
+ "app_name": "application-name"
}
```

### CODE REVIEW

File: lib/srv/app/session.go
Comment (r0mant) [inline lib/srv/app/session.go:169]: Let's include `AppURI` here and everywhere else too. This way, if the app is deleted and added with the same name but pointing to a different web application, you'll still be able tell them apart in the events.

File: lib/srv/app/transport.go
Comment (r0mant) [inline lib/srv/app/transport.go:294]: `t.c.uri` is the AppURI, not AppPublicAddr, right?

File: lib/srv/app/session.go
Comment (r0mant) [inline lib/srv/app/session.go:85]: Instead of passing `name`, `uri` and `publicAddr` and other app fields individually, how about just passing entire `app` and let transport get stuff it needs from it?

**Review by r0mant — COMMENTED** [review]
@gabrielcorado Congrats on your first PR! 🎉

Left a couple suggestions but overall the approach makes sense to me.

Also, could you please add some test coverage to the PR? In general, we require that every change includes some unit or integration tests so make sure things don't regress.

File: integration/app_integration_test.go
Comment (r0mant) [inline integration/app_integration_test.go:1114]: Instead of making this a part of `createAppSession`, I would honestly decouple it and make a separate method e.g. `ensureAuditEvent` or something that would accept the event it expects.

File: integration/app_integration_test.go
Comment (r0mant) [inline integration/app_integration_test.go:1133]: Let's compare entire event objects instead of individual fields. Would make the test more robust.

File: lib/srv/app/server_test.go
Comment (r0mant) [inline lib/srv/app/server_test.go:47]: Is this a new dependency? Can we use standard library instead?

File: lib/srv/app/server_test.go
Comment (r0mant) [inline lib/srv/app/server_test.go:459]: Let's compare objects instead of individual fields.

File: lib/srv/app/server_test.go
Comment (r0mant) [inline lib/srv/app/server_test.go:481]: Let's compare objects instead of individual fields.

File: lib/srv/app/transport.go
Comment (r0mant) [inline lib/srv/app/transport.go:68]: I think these can now just be replaced by

```go
if c.app == nil {
  return trace.BadParameter("app missing")
}
```

**Review by r0mant — COMMENTED** [review]
Good job on the tests! Left a couple more suggestions, should be lgtm once addressed.

File: lib/srv/app/server_test.go
Comment (gabrielcorado) [reply lib/srv/app/server_test.go:47]: Not new. It is being used by the project already. I'm using it because I based another test, but I don't see any issue using the standard library.

Comment (r0mant) [reply lib/srv/app/server_test.go:47]: @gabrielcorado If we're already using it, it's fine.

File: integration/app_integration_test.go
Comment (gabrielcorado) [reply integration/app_integration_test.go:1141]: I've isolated the search part, but the assertion is still inside this function since it is specific. I've made a function similar to the [`checkHTTPResponse`](https://github.com/gravitational/teleport/blob/master/lib/srv/app/server_test.go#L415).

File: api/types/events/events.proto
Comment (r0mant) [inline api/types/events/events.proto:1176]: You can mark this field as deprecated using GRPC tag:

```suggestion
    string PublicAddr = 7 [ (gogoproto.jsontag) = "public_addr", deprecated = true ];
```

It won't do anything meaningful, just update the generated godoc to say that it's deprecated, but still useful sometimes to convey intent.

File: integration/app_integration_test.go
Comment (r0mant) [reply integration/app_integration_test.go:1141]: Yeah the assertion is fine, but I was mostly wondering why `ensureAuditEvent` needs to be called as a part of `createAppSession`? Instead of checking for the event in effectively every app access test, can we have a dedicated test that specifically tests for expected audit log events?

For example, it could do:

1. `createAppSession` to init app session.
2. `ensureAuditEvent` to check that session.start event was emitted.
3. Make a request to the app.
4. `ensureAuditeEvent` again to check that session.chunk event was emitted.

Would be easier to troubleshoot and catch potential issues. And you also wouldn't need to update "createAppSession" function signature to pass extra 2 arguments. And as another benefit, no other app access test would potentially block for 500ms waiting for events. WDYT?

**Review by r0mant — COMMENTED** [review]
@smallinsky Can you take a look as well?

File: lib/srv/app/server_test.go
Comment (smallinsky) [reply lib/srv/app/server_test.go:47]: Nit: 
We are trying to keep import in following format where packages are split into 3  std/teleport/external sections: 
```
import (
  "path/filepath"
  "time"
  ...
  
  "github.com/gravitational/teleport/api/types"
  ...
  
  "go.uber.org/atomic"
  ... 
)
```

File: integration/app_integration_test.go
Comment (gabrielcorado) [reply integration/app_integration_test.go:1141]: That makes sense. I added this audit event assertion on the `createAppSession` since there seem to be multiple "forms" of apps, like being accessed by a leaf node and WS, so the idea was to guarantee that all of those were producing the correct Session start events. But having a case-specific test for audit tests sounds better to me also. I'm just not sure if it would require all of those app configurations or just one is enough.

Anyway, I'm going to update the tests.

File: lib/srv/app/server_test.go
Comment (gabrielcorado) [reply lib/srv/app/server_test.go:516]: I've made the change, but the `requestEventsReceived` block (L513) needs to be inside the `checkHTTPResponse`. If not, the event never gets recorded into the variable. I'm not 100% sure why, but it seems that because the `checkHTTPResponse` closes the server, and since this test is using the Sync recorder, the server cancels the emit event.

**Comment by russjones** [comment]
@gabrielcorado Can you rebase and merge all your commits into one before merging?

File: api/types/events/events.proto
Comment (r0mant) [inline api/types/events/events.proto:1175]:
```suggestion
    // DELETE IN 10.0: this information is also present on the AppMetadata, remove in
    // Teleport 10.
```

**Review by r0mant — APPROVED** [review]
lgtm but please also wait for approval from @smallinsky 

Also, don't forget to backport to `branch/v8` pls.

**Review by smallinsky — APPROVED** [review]

**Pull request merged**
